File: observer-pattern/weather-station/weather-station.py
from abc import ABC, abstractmethod

# Observer (interface)
class Observer(ABC):
  @abstractmethod
  def update(self, temp: int):
    """React to temperature updates"""
    raise NotImplementedError

# Concrete observer: Say a display device
class DisplayDevice(Observer):
  def update(self, temperature: int):
    print(f'Display device (observer): Temperature updated to {temperature} degrees')
    # We can trigger some code here to update the display value according to the value of temperature received.
    # print('Display pixels updated')


# Subject
class WeatherStation:
  def __init__(self):
    self._observers = []
    self._temperature = 0

  def add_observer(self, observer: Observer):
    self._observers.append(observer)

  def remove_observer(self, observer: Observer):
    self._observers.remove(observer)

  def notify_observers(self):
    """Notify all observers about the temperature change"""
    for obs in self._observers:
      obs.update(self._temperature)

  def set_temperature(self, temp: int):
    """Set the temperature and notify observers"""
    print(f'Setting temperature to {temp} degrees')
    self._temperature = temp
    self.notify_observers()



# Dry run
if __name__ == '__main__':
  station = WeatherStation()
  display_device = DisplayDevice()

  station.add_observer(display_device) # Register display device as an observer
  
  import time
  import random

  for i in range(3):
    sleep_time = random.randint(1, 3) * i

    station.set_temperature(14 * i + 3)

    print(f'[DEBUG]: Sleeping for {sleep_time} seconds \n\n')
    time.sleep(sleep_time)

File: observer-pattern/weather-station/weather-station.ts
interface Observer {
  update(temperature: number): void;
}

// Concrete observer
class DisplayDevice implements Observer {
  update(temperature: number): void {
    console.log(
      `Display (observer): Temperature updated to ${temperature} degrees.`
    );
  }
}

// Subject
class WeatherStation {
  private observers: Observer[] = [];
  private temperature = 0;

  addObserver(obs: Observer) {
    this.observers.push(obs);
  }

  removeObserver(obs: Observer) {
    this.observers = this.observers.filter((o) => o != obs);
  }

  notifyObservers() {
    for (let obs of this.observers) {
      obs.update(this.temperature);
    }
  }

  setTemperature(temp: number) {
    console.info(`Setting temperature to ${temp} deg.`);

    this.temperature = temp;
    this.notifyObservers();
  }
}

// Dry run
const station = new WeatherStation();
const displayDevice = new DisplayDevice();

station.addObserver(displayDevice); // Registering the display device as an observer.

for (let i = 0; i < 3; ++i) {
  station.setTemperature(31 * i + 5);
}
